pkg/virt-controller/watch/volume-migration/volume-migration.go

```diff
@@ -0,0 +1,336 @@
+	"kubevirt.io/kubevirt/pkg/controller"
+	storagetypes "kubevirt.io/kubevirt/pkg/storage/types"
+)
+
+const invalidUpdateErrMsg = "The volume can only be reverted to the previous version during the update"
+
+// invalidVols includes the invalid volumes for the volume migration
+type invalidVols struct {
+	hotplugged  []string
+	fs          []string
+	shareable   []string
+	luns        []string
+	dvs         []string
+	dvTemplates []string
+}
+
+func (vols *invalidVols) errorMessage() error {
+	var s strings.Builder
+	if len(vols.hotplugged) < 1 && len(vols.fs) < 1 &&
+		len(vols.shareable) < 1 && len(vols.luns) < 1 &&
+		len(vols.dvs) < 1 && len(vols.dvTemplates) < 1 {
+		return nil
+	}
+	s.WriteString("invalid volumes to update with migration:")
+	if len(vols.hotplugged) > 0 {
+		s.WriteString(fmt.Sprintf(" hotplugged: %v", vols.hotplugged))
+	}
+	if len(vols.fs) > 0 {
+		s.WriteString(fmt.Sprintf(" filesystems: %v", vols.fs))
+	}
+	if len(vols.shareable) > 0 {
+		s.WriteString(fmt.Sprintf(" shareable: %v", vols.shareable))
+	}
+	if len(vols.luns) > 0 {
+		s.WriteString(fmt.Sprintf(" luns: %v", vols.luns))
+	}
+	if len(vols.dvs) > 0 {
+		s.WriteString(fmt.Sprintf(" DVs: %v", vols.dvs))
+	}
+	if len(vols.dvTemplates) > 0 {
+		s.WriteString(fmt.Sprintf(" DV templates: %v", vols.dvTemplates))
+	}
+
+	return fmt.Errorf(s.String())
+}
+
+func updatedVolumesMapping(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) map[string]string {
+	updateVols := make(map[string]string)
+	vmVols := make(map[string]string)
+	// New volumes
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		if name := storagetypes.PVCNameFromVirtVolume(&v); name != "" {
+			vmVols[v.Name] = name
+		}
+	}
+	// Old volumes
+	for _, v := range vmi.Spec.Volumes {
+		name := storagetypes.PVCNameFromVirtVolume(&v)
+		if claim, ok := vmVols[v.Name]; ok && name != claim {
+			updateVols[v.Name] = claim
+		}
+	}
+	return updateVols
+}
+
+func ValidateVolumes(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) error {
+	var invalidVols invalidVols
+	if vmi == nil {
+		return fmt.Errorf("cannot validate the migrated volumes for an empty VMI")
+	}
+	if vm == nil {
+		return fmt.Errorf("cannot validate the migrated volumes for an empty VM")
+	}
+	updatedVols := updatedVolumesMapping(vmi, vm)
+	valid := true
+	disks := storagetypes.GetDisksByName(&vmi.Spec)
+	filesystems := storagetypes.GetFilesystemsFromVolumes(vmi)
+	dvTemplates := make(map[string]bool)
+	for _, t := range vm.Spec.DataVolumeTemplates {
+		dvTemplates[t.Name] = true
+	}
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		_, ok := updatedVols[v.Name]
+		if !ok {
+			continue
+		}
+
+		// Datavolumes should be transformed into PVCs
+		if v.VolumeSource.DataVolume != nil {
+			invalidVols.dvs = append(invalidVols.dvs, v.Name)
+			valid = false
+			continue
+		}
+		// The reference to the DV template needs to be removed
+		if _, ok := dvTemplates[v.Name]; ok {
```

Comment:
Don't think this is right dvtemplate name does not necessarily correspond to volume name